Mathlib/GroupTheory/SpecificGroups/PresentedDihedral.lean

```diff
@@ -0,0 +1,231 @@
+      (FreeGroup.of (@Generator.a n))) * (QuotientGroup.mk (FreeGroup.of Generator.b)))
+      (i : ℤ) (k : ℤ)
+    have h7 := @a_b_zpow_mem_normalClosure_eq_identity n
+    rw [h5] at h6
+    rw [← h6]
+    have h8 := zpow_add ((QuotientGroup.mk (s := Subgroup.normalClosure (Rels n))
+      (FreeGroup.of (@Generator.a n))) * (QuotientGroup.mk (FreeGroup.of Generator.b)))
+      (i + k - n : ℤ) (n : ℤ)
+    simp only [QuotientGroup.mk_mul] at h7
+    rw [h8, h7, mul_one]
+
+def GenMapInv : DihedralGroup n → PresentedDihedralGroup n
+  | r i => QuotientGroup.mk ((FreeGroup.of Generator.a * FreeGroup.of Generator.b) ^ (i : ℤ))
+  | sr i => QuotientGroup.mk ((FreeGroup.of Generator.a) *
+    (FreeGroup.of Generator.a * FreeGroup.of Generator.b) ^ (i - 1 : ℤ))
+
+lemma map_mul' : ∀ a b : DihedralGroup n, GenMapInv (a * b) = GenMapInv a * GenMapInv b := by
+  intro d₁ d₂
+  simp only [GenMapInv]
+  rcases d₁ with i | j
+  · rcases d₂ with k | l
+    · simp only [r_mul_r]
+      exact powLaw1
+    · simp only [r_mul_sr]
+      sorry
+      -- think I can figure the other sorry's as they will be similar to a certain extent
+  · rcases d₂ with k | l
+    · simp only [sr_mul_r]
+      sorry
+    · simp only [sr_mul_sr]
+      sorry
+
+/-- Homorphism from DihedralGroup n to group presentation of DihedralGroup n. -/
+def dihedralGroupHom : DihedralGroup n →* PresentedDihedralGroup n where
+  toFun := GenMapInv
+  map_one' := by
+    simp only [GenMapInv, one_def, QuotientGroup.mk_zpow, QuotientGroup.mk_mul,
+      ZMod.cast_zero, zpow_zero]
+  map_mul' := by
+    intro x y
+    simp only [GenMapInv]
+    exact map_mul' x y
+
+lemma dihedralGroupHom1 {i : ZMod n} : dihedralGroupHom (r i : DihedralGroup n) =
+    ((PresentedGroup.of Generator.a) * (PresentedGroup.of Generator.b)) ^ (i : ℤ) := by
+  unfold dihedralGroupHom
+  simp only [MonoidHom.coe_mk, OneHom.coe_mk, GenMapInv, PresentedGroup.of,
+    QuotientGroup.mk_zpow, QuotientGroup.mk_mul]
+
+lemma dihedralGroupHom2 {i : ZMod n} : dihedralGroupHom (sr i : DihedralGroup n) =
+    (PresentedGroup.of Generator.a) *
+    ((PresentedGroup.of Generator.a) * (PresentedGroup.of Generator.b)) ^ (i - 1 : ℤ) := by
+  unfold dihedralGroupHom
+  simp only [MonoidHom.coe_mk, OneHom.coe_mk, GenMapInv, PresentedGroup.of,
+    QuotientGroup.mk_zpow, QuotientGroup.mk_mul]
+
+theorem rightInverseHoms : Function.RightInverse (@dihedralGroupHom n) (@presentedGroupHom n) := by
+  intro dn
+  rcases dn with i | i
+  · rw [dihedralGroupHom1]
+    have h1 := @a_b_relsGenMapOf_pow n i
+    simp only [map_zpow, map_mul]
+    simp only [map_mul, ZMod.int_cast_cast, ZMod.cast_id', id_eq] at h1
+    assumption
+  · rw [dihedralGroupHom2]
+    have h1 := @a_a_b_relsGenMapOf_pow n i
+    simp only [map_mul, map_zpow]
+    simp only [map_mul, map_zpow, ZMod.int_cast_cast, ZMod.cast_id', id_eq] at h1
+    assumption
+
+theorem leftInverseHoms : Function.LeftInverse (@dihedralGroupHom n) (@presentedGroupHom n) := by
+  intro pd
```

Comment:
This is a bad move. Just check the claim on the generators rather than an arbitrary element of the presented group.

Reply:
This is the part I am not fully understanding.  Is there a rough sketch/example that you can give me and how it would plug in to other parts of the code?

Reply:
I am currently updating branch with the presentation that you mentioned.